_user/rotrectangle/main.c

```diff
@@ -0,0 +1,277 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Rotating Rectangle
+ *
+ * Example of user application
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+/*                 a
+*     (x0,y0) ...............
+*           |               |
+*           |      .        |
+*         b |   (xc,yc)     |
+*           |...............|
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <graph.h>
+#include <time.h>
+#include <unistd.h>
+#include <math.h>
+#include <termios.h>
+
+#include <sys/threads.h>
+
+
+typedef struct {
+	unsigned int xc;
+	unsigned int yc;
+	unsigned int a;
+	unsigned int b;
+	unsigned int color;
+} rectangle_t;
+
+
+typedef struct {
+	rectangle_t rec;
+	rectangle_t prev;
+	unsigned int speed;
+	unsigned int maxxc;
+	unsigned int maxyc;
+	unsigned int minxc;
+	unsigned int minyc;
+	char stop;
+} rotrectangle_t;
+
+
+/* Enables/disables canon mode and echo */
+static int rotrectangle_switchmode(int canon)
+{
+	struct termios state;
+
+	tcgetattr(STDIN_FILENO, &state);
+	if (canon) {
+		state.c_lflag |= ICANON;
+		state.c_lflag |= ECHO;
+	}
+	else {
+		state.c_lflag &= ~ICANON;
+		state.c_lflag &= ~ECHO;
+		state.c_cc[VMIN] = 1;
+	}
+	tcsetattr(STDIN_FILENO, TCSANOW, &state);
+
+	return 0;
+}
+
+
+static void rotrectangle_getkey(void *arg)
+{
+	char ch;
+	rotrectangle_t *self = (rotrectangle_t *)arg;
```

Comment:
Error: writing/reading the same struct from 2 threads which is not marked as `volatile` - may work just by accident. Please fix it.

There still might be glitches due to not guarding better the access to the struct, but from the current code I assume they would be minimal.